scripts/fetch_search_database.py

```diff
@@ -0,0 +1,338 @@
+import json
+import shutil
+import sqlite3
+import subprocess
+import time
+from pathlib import Path
+
+import click
+import requests
+import xmltodict
+
+API_URL = "https://www.barbershoptags.com/api.php"
+OUT_DIR = Path(__file__).resolve().parent.parent / "out"
+# We name this `.otf` so that GitHub Pages will transparently gzip it for us when serving it.
+# Specifically GitHub has said a few times they use a CDN to serve the content, eg:
+# https://github.blog/2014-01-07-faster-more-awesome-github-pages/
+# And some sources claim it's Fastly, though that's hard to verify. In any case, when the file
+# name is just "*.sqlite" setting `Accept-Encoding: gzip` doesn't result in a smaller file, whereas
+# with the `.otf` extension it's distinctly smaller over the wire. I picked `.otf` from this list of
+# extensions that would likely work with auto-compression (though again, it's unclear if it's actually
+# Fastly under the hood):
+# https://docs.fastly.com/en/guides/enabling-automatic-compression#setting-up-a-compression-policy
+SQL_NAME = "tags_db.sqlite.otf"
+MANIFEST_NAME = "manifest.json"
+PAGE_SIZE = 100
+# Bump this if the format changes
+SCHEMA_VERSION = 0
+
+
+def fetch_xml_batches() -> list[dict]:
+    print("About to start fetching batches")
+    batches = []
+    i = 0
+    while True:
+        response = requests.get(
+            API_URL,
+            params={
+                "Sortby": "posted",
+                "n": str(PAGE_SIZE),
+                "start": str(i * PAGE_SIZE + 1),
+            },
+        )
+
+        response.raise_for_status()
+        data = xmltodict.parse(
+            response.content.decode(encoding="utf-8", errors="backslashreplace")
+        )
+        batches.append(data)
+        print(
+            f"Fetched page {i} ({data['tags']['tag'][0]['id']} - {data['tags']['tag'][-1]['id']})"
+        )
+        if int(data["tags"]["@count"]) < PAGE_SIZE:
+            break
+
+        i += 1
+
+    return batches
+
+
+def massage_tag_row(tag: dict) -> dict:
+    keys = [
+        "id",
+        "Title",
+        "AltTitle",
+        "Arranger",
+        "WritKey",
+        "Lyrics",
+        "Collection",
+        "Parts",
+        "Posted",
+        "SheetMusicAlt",
+        "Quartet",
+        "QWebsite",
+    ]
+    for key in keys:
+        tag.setdefault(key, "")
+    return tag
+
+
+def parse_batches_to_tags(batches: list[dict]) -> list[dict]:
+    return [massage_tag_row(t) for batch in batches for t in batch["tags"]["tag"]]
+
+
+def prepare_out_dir(out_dir: Path) -> None:
+    # Make sure it exists and clear out everything in here (aside from, say, the .git dir)
+    out_dir.mkdir(exist_ok=True, parents=True)
+    for path in out_dir.iterdir():
+        if not path.name.startswith("."):
+            if path.is_dir():
+                shutil.rmtree(path)
+            else:
+                path.unlink()
```

Comment:
Ahhh yeah this needs more explanation. I'm not totally sure where to put it since it's a bit of complicated interplay between this file and the `.yml` workspace file. I've put a comment here and expanded the docstring on `main` to talk about it a bit more; does that help clear things up?